### PR TITLE
Add ExtraProgramPages app call field support

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

* Use ExtraProgramPages from app create txn in accounting
* Add TotalExtraAppPages and TotalAppSchema calculated fields to account response

## Test Plan

Added sh test for total schema and go unit test for extra pages

### CODE REVIEW

**Comment by tolikzinovyev** [comment]
could you make travis happy?

File: idb/postgres/postgres.go [hunk 1]
```diff
@@ -2222,10 +2233,22 @@ func (db *IndexerDb) yieldAccountsThread(req *getAccountsRequest) {
...
 			}
 			account.AppsLocalState = &aout
 		}
 
+		if totalSchema.NumByteSlice != 0 || totalSchema.NumUint != 0 {
+			account.AppsTotalSchema = &totalSchema
+		}
+		if totalExtraPages != 0 {
+			account.AppsTotalExtraPages = &totalExtraPages
+		}
```
Comment (winder) [inline idb/postgres/postgres.go:2250]: nit: move this into the `len(appParamIndexes) > 0` block.

File: idb/postgres/postgres.go [hunk 2]
```diff
@@ -2092,6 +2095,9 @@ func (db *IndexerDb) yieldAccountsThread(req *getAccountsRequest) {
 			*account.CreatedAssets = cal
 		}
 
+		var totalSchema models.ApplicationStateSchema
+		var totalExtraPages uint64
```
Comment (winder) [inline idb/postgres/postgres.go:2099]: nit: move this into the `len(appParamIndexes) > 0` block.

File: idb/postgres/postgres.go [hunk 3]
```diff
@@ -2222,10 +2233,22 @@ func (db *IndexerDb) yieldAccountsThread(req *getAccountsRequest) {
...
 				}
 				aout[i].KeyValue = ls[i].KeyValue.toModel()
+				if !*aout[i].Deleted {
+					totalSchema.NumByteSlice += ls[i].Schema.NumByteSlice
+					totalSchema.NumUint += ls[i].Schema.NumUint
+				}
+
 			}
 			account.AppsLocalState = &aout
 		}
```
Comment (winder) [inline idb/postgres/postgres.go:2244]: I think it's fine to show these even if they're 0. That would also make totalExtraPages testable with the existing dataset

**Comment by winder** [comment]
For testing, please enable this section in the validator and add the new `apps-total-extra-pages` field:
https://github.com/algorand/indexer/blob/develop/cmd/validator/struct_processor.go#L74

For automated tests, we don't have a great way to do it. You could modify the bash tests. There should be some data for the total schema, but you'd have to check the total pages equals 0:
https://github.com/algorand/indexer/blob/develop/test/common.sh

The last option would be to add a go integration test, this is pretty cumbersome though, so maybe the bash script is easier:
https://github.com/algorand/indexer/blob/develop/idb/postgres/postgres_integration_test.go#L788

File: idb/postgres/postgres.go (hunk 2 above)
Comment (algorandskiy) [reply idb/postgres/postgres.go]: Only `totalExtraPages` can be moved - done. `totalSchema` is calculated for both created and opted-in apps

**Comment by algorandskiy** [comment]
> could you make travis happy?

I guess after go sdk release and updating the reference.

**Comment by jasonpaulos** [comment]
@algorandskiy The swagger spec file also needs the extra pages field in `TransactionApplication` (line 1571).

**Comment by algorandskiy** [comment]
> @algorandskiy The swagger spec file also needs the extra pages field in `TransactionApplication` (line 1571).

Nice catch, thank you, I thought it shares `ApplicationParams` with apps but it does not.

**Review by tolikzinovyev — COMMENTED** [review]
I personally think the bash tests are very cumbersome, both for the writer and the reader. First you deal with a language that doesn't type check, and second you need to modify huge sql dump files.

With go tests you can start with a clean state and construct transactions you need in a convenient way. I only write tests in go.

File: idb/postgres/postgres.go [hunk 4]
```diff
@@ -2160,13 +2166,22 @@ func (db *IndexerDb) yieldAccountsThread(req *getAccountsRequest) {
 						NumUint:      apps[i].LocalStateSchema.NumUint,
 					}
 				}
+				if aout[outpos].Deleted != nil && !*aout[outpos].Deleted {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply idb/postgres/postgres.go:2169]: I think you are right, fixed

File: idb/idb.go [hunk 5]
```diff
@@ -452,6 +456,7 @@ type AppReverseDelta struct {
 	ClearStateProgram []byte                 `codec:"clearp,omitempty"`
 	LocalStateSchema  sdk_types.StateSchema  `codec:"lsch,omitempty"`
 	GlobalStateSchema sdk_types.StateSchema  `codec:"gsch,omitempty"`
+	ExtraProgramPages uint32                 `codec:"epp"`
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply idb/idb.go:459]: done

**Comment by algorandskiy** [comment]
Updated go sdk to 1.9.0

Can someone explain how to add [e2e_subs test](https://github.com/algorand/go-algorand/pull/2306/files#diff-081922fe8c3223cec073ced567b65001b91cfe3ae389ea7312e3034062fe0243R83) into test suite data?

**Comment by algorandskiy** [comment]
Thanks Tolik and Will for pointing to the test suite. Added extra pages test.

File: idb/idb.go [hunk 6]
```diff
@@ -452,6 +456,7 @@ type AppReverseDelta struct {
 	ClearStateProgram []byte                 `codec:"clearp,omitempty"`
 	LocalStateSchema  sdk_types.StateSchema  `codec:"lsch,omitempty"`
 	GlobalStateSchema sdk_types.StateSchema  `codec:"gsch,omitempty"`
+	ExtraProgramPages uint32                 `codec:"epp,omitempty"`
```
Comment (tolikzinovyev) [inline idb/idb.go:459]: We should probably set this field in `commitRoundAccounting()` or remove it. The situation with `AppReverseDelta` is strange. It's not used, not tested, and `LocalStateSchema` and `GlobalStateSchema` fields are completely unused. I really want to remove it. @winder what do you think about this?

File: idb/postgres/postgres.go [hunk 7]
```diff
@@ -2222,10 +2237,19 @@ func (db *IndexerDb) yieldAccountsThread(req *getAccountsRequest) {
...
+				if aout[i].Deleted == nil || !*aout[i].Deleted {
+					totalSchema.NumByteSlice += ls[i].Schema.NumByteSlice
+					totalSchema.NumUint += ls[i].Schema.NumUint
+				}
+
 			}
 			account.AppsLocalState = &aout
 		}
 
+		if totalSchema.NumByteSlice != 0 || totalSchema.NumUint != 0 {
```
Comment (tolikzinovyev) [inline idb/postgres/postgres.go:2249]: probably better to write `totalSchema != models.ApplicationStateSchema{}`

**Comment by tsachiherman** [comment]
> Updated go sdk to 1.9.0
> 
> Can someone explain how to add [e2e_subs test](https://github.com/algorand/go-algorand/pull/2306/files#diff-081922fe8c3223cec073ced567b65001b91cfe3ae389ea7312e3034062fe0243R83) into test suite data?

@brianolson - penny for your thoughts ?

File: idb/idb.go (hunk 6 above)
Comment (algorandskiy) [reply idb/idb.go:459]: It is set https://github.com/algorand/indexer/pull/545/files#diff-8fbec00be6e6b2f77a294cbe9a3ee8e3fbdcb5eb13be3046e5da8277302264c2R1140

File: idb/postgres/postgres_integration_test.go [hunk 8]
```diff
@@ -890,3 +890,58 @@ func TestAssetFreezeTxnParticipation(t *testing.T) {
...
+	row := db.db.QueryRow("SELECT params FROM app WHERE creator = $1", test.AccountA[:])
+
+	var paramsStr []byte
+	err = row.Scan(&paramsStr)
+	require.NoError(t, err)
+
+	var ap AppParams
+	err = encoding.DecodeJSON(paramsStr, &ap)
+	require.Equal(t, uint32(1), ap.ExtraProgramPages)
+}
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply idb/postgres/postgres_integration_test.go:947]: done

File: idb/postgres/postgres.go (hunk 7 above)
Comment (algorandskiy) [reply idb/postgres/postgres.go:2249]: done

File: idb/postgres/postgres_integration_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply idb/postgres/postgres_integration_test.go:965]: @tolikzinovyev any idea why does `GetAccounts` return `sql: no rows in result set` ?

**Comment by algorandskiy** [comment]
Removed `Applications` and `GetAccounts` tests for now - they require some additional setup and I can't figure out what. Will do add these in a separate PR.

File: idb/postgres/postgres_integration_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (winder) [reply idb/postgres/postgres_integration_test.go:949]: If you set a breakpoint you can grab the connection string and look in the DB. I'm going to pull the code and take a look as well

**Comment by winder** [comment]
> > Updated go sdk to 1.9.0
> > Can someone explain how to add [e2e_subs test](https://github.com/algorand/go-algorand/pull/2306/files#diff-081922fe8c3223cec073ced567b65001b91cfe3ae389ea7312e3034062fe0243R83) into test suite data?
> 
> @brianolson - penny for your thoughts ?

There is some documentation here, but it depends on deprecated features of indexer...
https://github.com/algorand/indexer/tree/develop/test#integration-test

File: idb/idb.go (hunk 6 above)
Comment (tolikzinovyev) [reply idb/idb.go:459]: That's `AppParams`, not `AppReverseDelta`. Or am I missing something?

**Comment by brianolson** [comment]
There's a small fix needed to misc/buildtestdata.sh; I'm working on running this now to re-prove that system.

File: idb/postgres/postgres_integration_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tolikzinovyev) [reply idb/postgres/postgres_integration_test.go:949]: If you show the code, I can take a look too.

**Comment by brianolson** [comment]
This makes buildtestdata.sh work,
but indexer develop branch fails the resulting data; hopefully the code changes in this PR fix that.

**Comment by winder** [comment]
@algorandskiy the no rounds issue is because we changed these helpers to lookup the round. To add this test I modified your code as follows:
```diff
                err := state.AddTransaction(&txnRow)
                require.NoError(t, err)
+
+               block := test.MakeBlockForTxns(test.Round, &txn)
+               blockImporter := importer.NewDBImporter(db)
+               txnCount, err := blockImporter.ImportDecodedBlock(&block)
+               assert.NoError(t, err, "failed to import")
+               assert.Equal(t, 1, txnCount)
        }
 
        err = db.CommitRoundAccounting(state.RoundUpdates, test.Round, &types.BlockHeader{})
@@ -946,4 +953,14 @@ func TestAppExtraPages(t *testing.T) {
        var ap AppParams
        err = encoding.DecodeJSON(paramsStr, &ap)
        require.Equal(t, uint32(1), ap.ExtraProgramPages)
+
+       rows, _ := db.GetAccounts(context.Background(), idb.AccountQueryOptions{EqualToAddress: test.AccountA[:]})
+       num := 0
+       for row := range rows {
+               require.NoError(t, row.Error)
+               num++
+               require.NotNil(t, row.Account.AppsTotalExtraPages, "we should have this field")
+               require.Equal(t, uint64(1), *row.Account.AppsTotalExtraPages)
+       }
+       require.Equal(t, 1, num)
 }
```

File: idb/idb.go (hunk 6 above)
Comment (algorandskiy) [reply idb/idb.go:459]: heh, I got confused. Added assignment to `reverseDelta` as well.

**Comment by algorandskiy** [comment]
Thank you Will, added `Applications` and `GetAccounts` tests back.

File: idb/postgres/postgres.go [hunk 9]
```diff
@@ -1136,6 +1137,9 @@ ON CONFLICT (addr, assetid) DO UPDATE SET amount = account_asset.amount + EXCLUD
 					return fmt.Errorf("app delta apply err r=%d i=%d app=%d, %v", adelta.Round, adelta.Intra, adelta.AppIndex, err)
 				}
 			}
+			reverseDelta.ExtraProgramPages = adelta.ExtraProgramPages
```
Comment (tolikzinovyev) [inline idb/postgres/postgres.go:1140]: I think it should be `reverseDelta.ExtraProgramPages = state.ExtraProgramPages`.

File: api/generated/v2/types.go [hunk 10]
```diff
@@ -605,6 +611,9 @@ type TransactionApplication struct {
 	// \[apsu\] Logic executed for application transactions with on-completion set to "clear". It can read and write global state for the application, as well as account-specific local state. Clear state programs cannot reject the transaction.
 	ClearStateProgram *[]byte `json:"clear-state-program,omitempty"`
 
+	// \[epp\] specifies the additional app program len requested in pages.
+	ExtraProgramPages *uint64 `json:"extra-program-pages,omitempty"`
```
Comment (tolikzinovyev) [inline api/generated/v2/types.go:615]: We need to set this field in `api/converter_utils.go`.

File: idb/postgres/postgres.go (hunk 9 above)
Comment (algorandskiy) [reply idb/postgres/postgres.go]: ops, fixed

File: api/generated/v2/types.go (hunk 10 above)
Comment (algorandskiy) [reply api/generated/v2/types.go:615]: nice catch

Comment (tolikzinovyev) [reply api/generated/v2/types.go:615]: Thanks! Could you also add a test for `api.LookupTransaction()`?

**Review by winder — APPROVED** [review]

**Review by brianolson — APPROVED** [review]

**Pull request merged**